### ISSUE TITLE
Aliases not working

### ISSUE THREAD

**Issue by johnf** [microsoft/rnx-kit#1910]
### What happened?

In my react project I use aliases to import from ```src``` e.g.

```typescript
import { Section } from './src/Section';
// versus
import { Section } from '@demo/Section';
```

This doesn't seem to be supported by rnx-start, I get the following error

```
Reloading app...
App.tsx:29:31 - error TS2307: Cannot find module '@demo/Section' or its corresponding type declarations.

29 import {Section, styles} from '@demo/Section';
                                 ~~~~~~~~~~~~~~~
```

I'm guessing I probably need a slight config change but haven't been able to work it out.

I'm happy to provide a PR for documentation if someone can point me in the right direction.

### Affected Package

@rnx-kit/cli

### Version

0.14.8

### Which platforms are you seeing this issue on?

- [X] Android
- [X] iOS
- [ ] macOS
- [ ] Windows

### System Information

```true
info Fetching system and libraries information...
System:
    OS: Linux 5.15 Ubuntu 20.04.5 LTS (Focal Fossa)
    CPU: (12) x64 Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
    Memory: 10.86 GB / 30.99 GB
    Shell: 5.0.17 - /bin/bash
  Binaries:
    Node: 16.15.0 - ~/.nvm/versions/node/v16.15.0/bin/node
    Yarn: 1.22.18 - ~/.nvm/versions/node/v16.15.0/bin/yarn
    npm: 8.5.5 - ~/.nvm/versions/node/v16.15.0/bin/npm
    Watchman: 4.9.0 - /usr/bin/watchman
  SDKs:
    Android SDK:
      API Levels: 30, 31, 32, 33
      Build Tools: 28.0.3, 29.0.2, 29.0.3, 30.0.2, 30.0.3, 31.0.0, 31.0.0, 32.1.0, 33.0.0, 33.0.0
      System Images: android-30 | Google APIs Intel x86 Atom, android-30 | Google Play Intel x86 Atom, android-31 | Google APIs Intel x86 Atom_64, android-33 | Google APIs Intel x86 Atom_64
      Android NDK: Not Found
  IDEs:
    Android Studio: AI-213.7172.25.2113.9014738
  Languages:
    Java: 11.0.16 - /usr/bin/javac
  npmPackages:
    @react-native-community/cli: Not Found
    react: 18.1.0 => 18.1.0 
    react-native: 0.70.1 => 0.70.1 
  npmGlobalPackages:
    *react-native*: Not Found
```

### Steps to Reproduce

I've provided a demo repo at https://github.com/johnf/rnk-kit-aliases-demo

Simply ```yarn start```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by tido64** [comment]
Hi @johnf, I had a look at your repro and it looks like it bundled successfully despite the error:

![image](https://user-images.githubusercontent.com/4123478/193283420-9fe6f243-a34d-4229-810e-1d0e8e277910.png)

The error looks like it comes from TypeScript, which isn't using Babel, so it will naturally stumble upon your alias. You _could_ disable TypeScript type checking in the dev server like below:

```diff
diff --git a/package.json b/package.json
index 0bf6e96..e6d672c 100644
--- a/package.json
+++ b/package.json
@@ -45,5 +45,10 @@
       "json",
       "node"
     ]
+  },
+  "rnx-kit": {
+    "server": {
+      "typescriptValidation": false
+    }
   }
 }
```

Regardless, I would advise against using `babel-plugin-module-resolver` if you care about being able to run `tsc`.

cc @afoxman

**Comment by johnf** [comment]
Hi @tido64 

You are right it is only the typescript validation that is the issue (one of the main reasons I want to switch is so metro tells me about typescript errors).

In my tsconfig.json I have
```json
{
  "baseUrl": ".",
    "paths": {
      "*": ["src/*"],
      "@gladly/*": ["src/*/index", "src/*"]
    },
}
```
If I use ```tsc```, typescript resolves the aliases fine.

I've dug into the code and my guess is the problem lies in ```typescript-react-native-resolver```, I think that it is replacing the typescript resolver entirely and doesn't support ```paths``` like Typescript does.

**Comment by itsramiel** [comment]
Hey, I am stumbled across the same issue. I am trying to detect circular dependencies, but getting the errors about not finding module.

The fix to add 
```
"rnx-kit": {
		"server": {
			"typescriptValidation": false
		}
	},
```

Did not work and still showed the error message

**Comment by tido64** [comment]
> Hi @tido64
> 
> You are right it is only the typescript validation that is the issue (one of the main reasons I want to switch is so metro tells me about typescript errors).
> 
> In my tsconfig.json I have
> 
> ```json
> {
>   "baseUrl": ".",
>     "paths": {
>       "*": ["src/*"],
>       "@demo/*": ["src/*/index", "src/*"]
>     },
> }
> ```
> 
> If I use `tsc`, typescript resolves the aliases fine.
> 
> I've dug into the code and my guess is the problem lies in `typescript-react-native-resolver`, I think that it is replacing the typescript resolver entirely and doesn't support `paths` like Typescript does.

@afoxman: Does our resolver support `paths`? Is this something you can look into?

**Comment by tido64** [comment]
> Hey, I am stumbled across the same issue. I am trying to detect circular dependencies, but getting the errors about not finding module.
> 
> The fix to add
> 
> ```
> "rnx-kit": {
> 		"server": {
> 			"typescriptValidation": false
> 		}
> 	},
> ```
> 
> Did not work and still showed the error message

Which version of the CLI are you using? We changed the config schema 0.14. If you are on an older version, this config won't work. Also, can you please file a separate issue?

**Comment by afoxman** [comment]
The rnx-kit CLI (and `typescript-react-native-resolver`) doesn't support TypeScript's path remapping system -- e.g. `baseUrl` and `paths`. ☹️ 

Unfortunately, that code is really complicated, so replicating it will be hard to do and get right. TS doesn't offer an API-level call that lets you "remap" a path either.

The good news is that with TypeScript 4.7, rnx-kit doesn't need to overwrite the resolver with its own. TS natively supports RN platform-specific resolution (`moduleSuffixes`).

I think the path forward here is to update rnx-kit with two changes:

1. When running against TS < 4.7, fail when `baseUrl` or `paths` is set in any tsconfig. Explain the issue and a path forward by upgrading to TS 4.7. This will save users time and energy debugging.
2. When running against TS >= 4.7, don't intercept and replace the resolver. Instead, program it using `moduleSuffixes`.

I'll open two bugs and link them together AND to this issue.

**Comment by johnf** [comment]
@afoxman That sounds perfect. Thanks!

**Comment by afoxman** [comment]
In the meantime, the workaround here is to disable `typescriptValidation`.

If you want TypeScript to check your code while you're waiting on these fixes (no promises on timeline), you can run `tsc` as pre-bundling step. Make sure you are using TS 4.7. You'll need to set `moduleSuffixes`, and likely have one tsconfig per platform you want to validate.

e.g. 
tsconfig.ios.json: `{ ... , moduleSuffixes = ['ios', 'native', ''], ...}`
tsconfig.android.json: `{ ..., moduleSuffixes = ['android', 'native', ''], ...}`

**Comment by johnf** [comment]
@afoxman by pre-bundling step do you mean adding something before the start command in the scripts section of packae.json?
Or does metro support a pre-bundling hook? I did some quick googling but couldn't find anything relevant.

**Comment by afoxman** [comment]
> @afoxman by pre-bundling step do you mean adding something before the start command in the scripts section of packae.json? Or does metro support a pre-bundling hook? I did some quick googling but couldn't find anything relevant.

I meant running before rnx-start, so that you get type-checking before trying to run the app.

**Comment by johnf** [comment]
> * When running against TS < 4.7, fail when `baseUrl` or `paths` is set in any tsconfig. Explain the issue and a path forward by upgrading to TS 4.7. This will save users time and energy debugging. See issue [rnx-kit CLI should fail when it encounters tsconfig with path-remapping (TS < 4.7) #1944](https://github.com/microsoft/rnx-kit/issues/1944).
> * When running against TS >= 4.7, don't intercept and replace the resolver. Instead, program it using `moduleSuffixes`. You'll get the full TS resolver, including path-remapping support. See issue [rnx-kit should use `moduleSuffixes` when working against TS 4.7 or newer #1945](https://github.com/microsoft/rnx-kit/issues/1945).

I'm willing to have a go at creating a PR for both of these issues if you can provide some pointers

**Comment by tido64** [comment]
> The rnx-kit CLI (and `typescript-react-native-resolver`) doesn't support TypeScript's path remapping system -- e.g. `baseUrl` and `paths`. ☹️
> 
> Unfortunately, that code is really complicated, so replicating it will be hard to do and get right. TS doesn't offer an API-level call that lets you "remap" a path either.
> 
> The good news is that with TypeScript 4.7, rnx-kit doesn't need to overwrite the resolver with its own. TS natively supports RN platform-specific resolution (`moduleSuffixes`).
> 
> I think the path forward here is to update rnx-kit with two changes:
> 
> 1. When running against TS < 4.7, fail when `baseUrl` or `paths` is set in any tsconfig. Explain the issue and a path forward by upgrading to TS 4.7. This will save users time and energy debugging. See issue [rnx-kit CLI should fail when it encounters tsconfig with path-remapping (TS < 4.7) #1944](https://github.com/microsoft/rnx-kit/issues/1944).
> 2. When running against TS >= 4.7, don't intercept and replace the resolver. Instead, program it using `moduleSuffixes`. You'll get the full TS resolver, including path-remapping support. See issue [rnx-kit should use `moduleSuffixes` when working against TS 4.7 or newer #1945](https://github.com/microsoft/rnx-kit/issues/1945).
> 
> I'll open two bugs and link them together AND to this issue.

fyi @titozzz 👆

**Comment by johnf** [comment]
@foxman See PR in #1956

Two things I could use guidance on.

The original resolveNames does some stuff around extensions, do we need to replicate this in some way?

We need to check if the user already has moduleSuffixes set.
If so would we just append our suffixes or do we need to be smarter?

**Comment by afoxman** [comment]
> The original resolveNames does some stuff around extensions, do we need to replicate this in some way?
No, you don't need to do this. TS already has similar logic, though it is expressed differently because their resolver functions handle more complex scenarios.

> We need to check if the user already has moduleSuffixes set. If so would we just append our suffixes or do we need to be smarter?
This is a complex problem. I'll walk you through my thinking.

_React Native_ packages using `moduleSuffixes` _could_ have multiple tsconfigs, each with a `moduleSuffixes` list per platform. It doesn't have to be this way, though. Regardless how they are set up, there's no mapping of platform to tsconfig file, so we can't find the "right" tsconfig.

This is why rnx-kit always reads "tsconfig.json" when type-checking.

If there are existing `moduleSuffixes` in there, we either overwrite, fail, or see if we show up in the list and only proceed if we do.

I like the 3rd option best. If we want to add "ios", "native" and "", then those three entries need to be in the list, in that order, even if there are other entries in/around them. If that is true, proceed quietly. If not, fail with a message explaining, in detail, what the incompatibility is.

This is where I get hung up, because the user could now be stuck. They have this error message and now way to proceed, other than to disable type checking, which is no good. Maybe this is where we start, and see what bug reports come in, so we can expand how it works once we get more perspective. (The error should include an "open a bug" link so they'll be likely to report what they are seeing).

**Comment by tido64** [comment]
Would it make sense to add a `project` option to let people specify which `tsconfig` file to use? `tsc` has `--project` flag for this purpose: https://www.typescriptlang.org/docs/handbook/compiler-options.html

**Comment by afoxman** [comment]
> Would it make sense to add a `project` option to let people specify which `tsconfig` file to use? `tsc` has a `--project` flag for this purpose: https://www.typescriptlang.org/docs/handbook/compiler-options.html

It could vary per dependency, so one flag won't be enough. Each RN package could have a different filename for "ios".

**Comment by tido64** [comment]
Wait, is that what `tsc` does as well? Does it read `tsconfig.json` of every dependency? I thought only the current project's `tsconfig.json` was read, otherwise you can quickly get into weird scenarios where some projects transpile to CJS and others to ESM. Not to mention the ordering of `moduleSuffixes` could wildly differ between dependencies.

**Comment by afoxman** [comment]
This is a good point. Our code reads each project's tsconfig because the rules vary per-package. 

Maybe this approach is wrong? We should instead ignore dependency tsconfig files and instead look only at the .d.ts files, which are the result of compiling with those rules?

**Comment by tido64** [comment]
> We should instead ignore dependency tsconfig files and instead look only at the .d.ts files, which are the result of compiling with those rules?

I think this is the right approach. We should not apply local `tsconfig.json` files as they can directly conflict with the current project's config. I would have a look at what `tsc` does to be sure, but I wonder if `.ts` files is a valid source if they are present. It probably would prefer `.d.ts` since they should be faster to parse?

**Comment by johnf** [comment]
> I like the 3rd option best. If we want to add "ios", "native" and "", then those three entries need to be in the list, in that order, even if there are other entries in/around them. If that is true, proceed quietly. If not, fail with a message explaining, in detail, what the incompatibility is.

@afoxman I think what you mean here is that we leave it up to the user to include these and if they are missing we throw an error.

I don't think this works though as CRA creates a single ```tsconfig.json```, which is used by ```android``` and ```ios```. It's only at metro time based on command line arguments or the URL that is requested that you know the platform.

So I think it makes more sense for us to inject them. I'd suggest that when ```tsconfig,json```
* has no moduleSuffixes: we add it based on the platform
* moduleSuffixes exists but contains none of what we are going to insert then we append our version

Alternatively, we could ask them to put some sort of placeholder like "RNX_PLATFORM_SUFFIXES" into moduleSuffixes and we use that to insert.

To be honest, I'm not sure what folks would be using moduleSiffuxes for, and I suspect it's going to be fairly rare your average user will be using it in a react native project?

**Comment by johnf** [comment]
> The original resolveNames does some stuff around extensions, do we need to replicate this in some way?

@afoxman also I'm still not sure about this bit

**Comment by afoxman** [comment]
> > No, you don't need to do this. TS already has similar logic, though it is expressed differently because their resolver functions handle more complex scenarios.

My answer got pulled into the reply-block :) 

No, you don't need to do this. TS already has similar logic, though it is expressed differently because their resolver functions handle more complex scenarios.

**Comment by afoxman** [comment]
> > I like the 3rd option best. If we want to add "ios", "native" and "", then those three entries need to be in the list, in that order, even if there are other entries in/around them. If that is true, proceed quietly. If not, fail with a message explaining, in detail, what the incompatibility is.
> 
> I don't think this works though as CRA creates a single `tsconfig.json`, which is used by `android` and `ios`. It's only at metro time based on command line arguments or the URL that is requested that you know the platform.

Does CRA use `moduleSuffixes` right now? I doubt it, because they're run into the need for separate TSConfigs, unless they put both "android" and "ios" into a single tsconfig (which wouldn't make sense).

> So I think it makes more sense for us to inject them. I'd suggest that when `tsconfig,json`
> 
> * has no moduleSuffixes: we add it based on the platform
> * moduleSuffixes exists but contains none of what we are going to insert then we append our version
> 
> Alternatively, we could ask them to put some sort of placeholder like "RNX_PLATFORM_SUFFIXES" into moduleSuffixes and we use that to insert.
> 
> To be honest, I'm not sure what folks would be using moduleSiffuxes for, and I suspect it's going to be fairly rare your average user will be using it in a react native project?

I don't like the idea of making these runtime changes since we don't know when `moduleSuffixes` would be used for RN or other purposes. We need to support `moduleSuffixes` as the general feature that it is, not an RN-specific TS addon. We won't always be able make it work, so we might have to fail with an informative message and let the developer figure it out (or submit a bug/PR we can work from). And I agree it will be unlikely, so this won't happen often and it won't break users often.

**Comment by johnf** [comment]
@afoxman OK makes sense. So for now I'll inject moduleSuffixes if it doesn't exist. If it does exist, we'll throw an exception with instructions pointing at a new bug I'll create.

**Comment by afoxman** [comment]
> @afoxman OK makes sense. So for now I'll inject moduleSuffixes if it doesn't exist. If it does exist, we'll throw an exception with instructions pointing at a new bug I'll create.

I'd prefer we check to see if the suffixes we want to add are already in the list, in the order we want. If not, then you can throw.

**Comment by johnf** [comment]
> I'd prefer we check to see if the suffixes we want to add are already in the list, in the order we want. If not, then you can throw.

Good call will do.

**Comment by tido64** [comment]
I'm going to close this issue as all the PRs mentioned in https://github.com/microsoft/rnx-kit/pull/1956#issuecomment-1336288319 have been merged and published. [@rnx-kit/cli@0.15.3](https://github.com/microsoft/rnx-kit/releases/tag/%40rnx-kit%2Fcli%400.15.3) should include all the latest changes. Please feel free to file a new issue if it doesn't work for you.

**Comment by johnf** [comment]
I've tested this again in the released version and it works perfectly. I was able to clear up a bunch of TS errors.
Thanks @afoxman  and @tido64!
